src/rust/collections/concurrent_ring.rs

```diff
@@ -0,0 +1,509 @@
+                // We know that the pop_offset did not move because it was not pointing at a valid message.
+                let old_len: usize = self.write_header(pop_offset, bytes);
+                assert_eq!(old_len, 0);
+                return Err(Fail::new(libc::EINVAL, "Buffer is too small to hold next message"));
+            },
+        };
+
+        let buf_ptr: *mut u8 = buf.as_mut_ptr();
+        unsafe {
+            // Pointer to beginning of valid data for this popped message.
+            let pop_ptr: *const u8 = self.buffer.get()[pop_offset + HEADER_SIZE] as *const u8;
+            // Copy the data from the ring buffer.
+            copy(pop_ptr, buf_ptr, pop_len - HEADER_SIZE)
+        };
+
+        // Move to next buffer.
+        self.free_space(pop_offset, pop_len);
+        Ok(pop_len - HEADER_SIZE)
+    }
+
+    /// Removes the next message from the ring buffer up to [len] bytes and copies into [buf]. This function may block
+    /// (spin).
+    #[allow(unused)]
+    pub fn pop(&self, buf: &mut [u8]) -> usize {
+        loop {
+            if let Ok(pop_len) = self.try_pop(buf) {
+                break pop_len;
+            }
+        }
+    }
+
+    /// Atomically writes a header at the indicated offset and returns the previous one.
+    fn write_header(&self, offset: usize, val: usize) -> usize {
+        let buffer_ptr: *mut u8 = unsafe { self.buffer.get_mut() }.as_mut_ptr();
+        let header_ptr: *mut u16 = unsafe { buffer_ptr.add(offset) } as *mut u16;
+        let header: &AtomicU16 = AtomicU16::from_mut(unsafe { &mut *header_ptr });
+        header.swap(val as u16, atomic::Ordering::Relaxed) as usize
+    }
+
+    /// Given a [push_offset] and [pop_offset] into the ring buffer, return available space for writing data. Always
+    /// leave one [HEADER_SIZE] space for distinguishing a full from empty buffer.
+    fn available_space(&self, push_offset: usize, pop_offset: usize) -> usize {
+        let used_space: usize = (push_offset + self.capacity() - pop_offset) % self.capacity();
+        self.capacity() - used_space - HEADER_SIZE
+    }
+
+    /// Align to u16 for the header offset.
+    fn align_header(offset: usize) -> usize {
+        // Round up to u16 aligned.
+        match offset % HEADER_SIZE {
+            0 => offset,
+            x => offset - x + HEADER_SIZE,
+        }
+    }
+
+    /// Reserves [len] bytes from the ring buffer. If successful, returns the offset of the beginning of the buffer,
+    /// else returns None.
+    fn allocate_space(&self, len: usize) -> Option<usize> {
+        let len_: usize = Self::align_header(len);
+        let push_offset: usize = peek(self.push_offset);
+        let pop_offset: usize = peek(self.pop_offset);
+
+        if len_ > self.available_space(push_offset, pop_offset) {
+            return None;
+        }
+        let new_offset: usize = (push_offset + len_) % self.capacity();
+        assert_ne!(new_offset, pop_offset);
```

Comment:
This is multi-process, so isn't it possible that some interleaved pops() happen while retrieving push and pop offsets? How is that coupling with concurrency?

Reply:
Because we write 0 into the header and do not move the pop_offset, that effectively locks the pop end of the ring because all other pops will block on the 0 length, which looks like there is nothing to pop.